### PR TITLE
[HLSL] Add implicit resource element type concepts to AST

### PR DESCRIPTION
**Pull request by bob80905**
This PR is step one on the journey to implement resource element type validation via C++20 concepts. The PR sets up the infrastructure for injecting implicit concept decls / concept specialization expressions into the AST, which will then be evaluated after template arguments are instantiated. This is not meant to be a complete implementation of the desired validation for HLSL,
there are a couple of missing elements:
1. We need the __builtin_hlsl_is_typed_resource_element_compatible builtin to be implemented.
2. We need other constraints, like is_intangible
3. We need to put the first 2 points together, and construct a finalized constraint expression, which should differ between typed and raw buffers

This is just an initial PR that puts some of the core infrastructure in place.

### CODE REVIEW

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 1]
```diff
@@ -483,10 +554,109 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
...
+  Expr *ConstraintExpr = getTypedBufferConstraintExpr(S, DeclLoc, T);
+
+  // Create a ConceptDecl
+  clang::ConceptDecl *conceptDecl = clang::ConceptDecl::Create(
+      context,
+      context.getTranslationUnitDecl(), // DeclContext
+      DeclLoc,                          // Source location of start of concept
+      DeclName,                         // Source location of end of concept
+      ConceptParams,                    // Template type parameter
+      ConstraintExpr                    // Expression defining the concept
```
[the remark this answers is not in the document]
Comment (damyanp) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:644]: Looks like the comment next to `DeclName` is incorrect though?

[the remark this answers is not in the document]
Comment (bob80905) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:644]: This is no longer the latest version of the file. But the reason I disagreed is because, the Create function does not take that many arguments. An extra argument was added when it shouldn't have been. I've removed the comments.

[the remark this answers is not in the document]
Comment (damyanp) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp]: Yeah, it spotted something questionable, but the suggested fix was worse!

File: clang/lib/AST/DeclTemplate.cpp [hunk 2]
```diff
@@ -1078,10 +1078,10 @@ ConceptDecl *ConceptDecl::Create(ASTContext &C, DeclContext *DC,
                                  TemplateParameterList *Params,
                                  Expr *ConstraintExpr) {
   bool Invalid = AdoptTemplateParameterList(Params, DC);
-  auto *TD = new (C, DC) ConceptDecl(DC, L, Name, Params, ConstraintExpr);
+  auto *CD = new (C, DC) ConceptDecl(DC, L, Name, Params, ConstraintExpr);
   if (Invalid)
-    TD->setInvalidDecl();
-  return TD;
+    CD->setInvalidDecl();
+  return CD;
```
Comment (damyanp) [inline clang/lib/AST/DeclTemplate.cpp:1084]: This seems like a NFC change in DeclTemplate?  Might be better to pull this out into a separate PR.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 3]
```diff
@@ -323,30 +324,99 @@ struct TemplateParameterListBuilder {
...
-        /* ParameterPack */ false);
+        /* Typename */ true,
+        /* ParameterPack */ false,
+        /* HasTypeConstraint*/ false);
     if (!DefaultValue.isNull())
       Decl->setDefaultArgument(
           S.Context, S.getTrivialTemplateArgumentLoc(DefaultValue, QualType(),
                                                      SourceLocation()));
-
+    Decl->setReferenced();
```
Comment (damyanp) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:334]: Is this a good bugfix we'd want independently of the concepts change? Or is this something that had to change because we're also using concepts now?

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 4]
```diff
@@ -323,30 +324,99 @@ struct TemplateParameterListBuilder {
...
     Params.emplace_back(Decl);
     return *this;
   }
 
-  BuiltinTypeDeclBuilder &finalizeTemplateArgs() {
+  ConceptSpecializationExpr *getConceptSpecializationExpr(Sema &S,
+                                                          ConceptDecl *CD) {
+    ASTContext &context = S.getASTContext();
+    SourceLocation loc = Builder.Record->getBeginLoc();
```
Comment (damyanp) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:342]:
```suggestion
    ASTContext &Context = S.getASTContext();
    SourceLocation Loc = Builder.Record->getBeginLoc();
```

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 5]
```diff
@@ -323,30 +324,99 @@ struct TemplateParameterListBuilder {
...
+    DeclarationNameInfo DNI(CD->getDeclName(), loc);
+    NestedNameSpecifierLoc NNSLoc;
+    DeclContext *DC = Builder.Record->getDeclContext();
+    TemplateArgumentListInfo TALI(loc, loc);
+
+    // assume that the concept decl has just one template parameter
```
Comment (damyanp) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:348]: Is this assumption something we capture in a simple assert?

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 6]
```diff
@@ -323,30 +324,99 @@ struct TemplateParameterListBuilder {
...
+    TemplateTypeParmDecl *ConceptTTPD = dyn_cast<TemplateTypeParmDecl>(
+        CD->getTemplateParameters()->getParam(0));
+
+    clang::TemplateTypeParmDecl *T = clang::TemplateTypeParmDecl::Create(
```
Comment (damyanp) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:352]: I wonder if some psuedo-c++ code in a comment showing what this these AST nodes represent would help someone reading this?

File: clang/test/AST/HLSL/StructuredBuffer-AST.hlsl [hunk 7]
```diff
@@ -1,64 +1,64 @@
-// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.0-library -x hlsl -ast-dump -DEMPTY %s | FileCheck -check-prefix=EMPTY %s
```
Comment (damyanp) [inline clang/test/AST/HLSL/StructuredBuffer-AST.hlsl:1]: hmm....maybe a line endings change issue on this file?  I'm wondering if some config in the root of this repo has changed since I've heard other people mention line endings getting messed up?

**Review by damyanp — COMMENTED** [review]
Some comments based on a quick read through:

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 8]
```diff
@@ -356,6 +426,9 @@ struct TemplateParameterListBuilder {
     QualType T = Builder.Template->getInjectedClassNameSpecialization();
     T = S.Context.getInjectedClassNameType(Builder.Record, T);
 
+    ArrayRef<TemplateArgument> TempArgs =
+        Builder.Template->getInjectedTemplateArgs();
```
Comment (pow2clk) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:430]: Unused variable `TempArgs`?

File: clang/lib/Sema/HLSLExternalSemaSource.cpp (hunk 3 above)
Comment (bob80905) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:334]: This is an odd addition that I am not 100% convinced is absolutely necessary, although I haven't tried testing without it. I don't think it's a bugfix per se. The reason this line exists is because I am trying to match this AST as perfectly as possible, and it served as a guide throughout my development (Thanks to Chris):
https://godbolt.org/z/Gq4hdTcrK
The AST that is constructed in this PR is constructed *as if* the code in the godbolt link were defined and parsed somewhere, and the code in the link uses typename and is referenced. So, I set the appropriate decl as referenced, and make sure it's represented with a typename rather than a class.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 9]
```diff
@@ -483,10 +555,106 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
...
+  T->setDeclContext(DC);
+  T->setReferenced();
+
+  // Create and Attach Template Parameter List to ConceptDecl
+  std::vector<NamedDecl *> TemplateParamsVec = {T};
+  llvm::ArrayRef<NamedDecl *> TemplateParams(TemplateParamsVec);
+
+  clang::TemplateParameterList *ConceptParams =
+      clang::TemplateParameterList::Create(context, DeclLoc, DeclLoc,
+                                           TemplateParams, DeclLoc, nullptr);
```
Comment (llvm-beanz) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:633]:
```suggestion
  clang::TemplateParameterList *ConceptParams =
      clang::TemplateParameterList::Create(context, DeclLoc, DeclLoc,
                                           {T}, DeclLoc, nullptr);
```

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 10]
```diff
@@ -323,30 +324,101 @@ struct TemplateParameterListBuilder {
...
+    TemplateArgument CSETA = TemplateArgument(CSETType);
+
+    std::vector<TemplateArgument> CSEConvertedArgsVec = {CSETA};
+    ArrayRef<TemplateArgument> CSEConvertedArgs = CSEConvertedArgsVec;
+
+    ImplicitConceptSpecializationDecl *ImplicitCSEDecl =
+        ImplicitConceptSpecializationDecl::Create(
+            context, Builder.Record->getDeclContext(), loc, CSEConvertedArgs);
+
+    const ConstraintSatisfaction CS(CD, ConceptConvertedArgs);
```
Comment (llvm-beanz) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:385]:
```suggestion
    clang::QualType CSETType = context.getTypeDeclType(T);

    TemplateArgument CSETA = TemplateArgument(CSETType);

    ImplicitConceptSpecializationDecl *ImplicitCSEDecl =
        ImplicitConceptSpecializationDecl::Create(
            context, Builder.Record->getDeclContext(), loc, {CSETA});

    const ConstraintSatisfaction CS(CD, {ConceptTA});
```

File: clang/lib/Sema/HLSLExternalSemaSource.cpp (hunk 6 above)
Comment (bob80905) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:352]: I've added some helpful comments.

File: clang/test/AST/HLSL/StructuredBuffer-AST.hlsl (hunk 7 above)
Comment (bob80905) [reply clang/test/AST/HLSL/StructuredBuffer-AST.hlsl:1]: Yes, line endings seem to be messed up. I've checked out main for both these files and only changed the necessary parts,  and pushed, but there must be some automatic thing my editor is doing to mess up the line endings.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp (hunk 3 above)
Comment (damyanp) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:334]: @llvm-beanz  / @bogner - did you write this originally?  Do you have any insights on what the right thing to do here is?

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 11]
```diff
@@ -323,30 +324,119 @@ struct TemplateParameterListBuilder {
...
     Params.emplace_back(Decl);
     return *this;
   }
 
-  BuiltinTypeDeclBuilder &finalizeTemplateArgs() {
+  /*
+  The desired result AST after this function constructs the concept
+  specialization Expression is as follows:
+
+  |-ConceptSpecializationExpr 0xdd5d408 <col:42, col:75> 'bool' Concept
```
Comment (damyanp) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:343]: Did some autoformatter do something bad with the formatting here?

File: clang/lib/Sema/HLSLExternalSemaSource.cpp (hunk 6 above)
Comment (damyanp) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:371]: Would it not be meaningful to show the C++ code that generated this AST?  I think that would help me understand it all better, but maybe if I saw it I'd realize I was wrong! :)

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 12]
```diff
@@ -483,10 +573,103 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
       .addDefaultHandleConstructor(S, RC);
 }
 
+BinaryOperator *getSizeOfLEQ16Expr(clang::ASTContext &context,
+                                   SourceLocation NameLoc,
+                                   TemplateTypeParmDecl *T) {
+  // Obtain the QualType for 'unsigned long'
+  clang::QualType unsignedLongType = context.UnsignedLongTy;
```
Comment (damyanp) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:580]: There's a `using namespace clang` at the beginning of this cpp file, so I don't think any of these `clang::` qualifications are necessary. A quick survey of the rest of the existing code suggests that they are usually omitted.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 13]
```diff
@@ -483,10 +573,103 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
       .addDefaultHandleConstructor(S, RC);
 }
 
+BinaryOperator *getSizeOfLEQ16Expr(clang::ASTContext &context,
```
Comment (damyanp) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:576]:
```suggestion
BinaryOperator *getSizeOfLEQ16Expr(clang::ASTContext &Context,
```

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 14]
```diff
@@ -483,10 +573,103 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
...
+
+  clang::UnaryExprOrTypeTraitExpr *sizeOfExpr = new (context)
+      clang::UnaryExprOrTypeTraitExpr(clang::UETT_SizeOf, TTypeSourceInfo,
+                                      unsignedLongType, NameLoc, NameLoc);
+
+  // Create an IntegerLiteral for the value '16' with size type
+  clang::QualType sizeType = context.getSizeType();
+  llvm::APInt sizeValue = llvm::APInt(context.getTypeSize(sizeType), 16);
+  clang::IntegerLiteral *sizeLiteral = new (context)
+      clang::IntegerLiteral(context, sizeValue, sizeType, NameLoc);
```
Comment (damyanp) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:597]: These variables don't follow the expected naming convention (of starting with a capital letter).

* `sizeOfExpr`
* `sizeType`
* `sizeLiteral`

Comment (damyanp) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:597]: There's many cases of it throughout these changes, I'm not going to call them all out now - but LMK if you think you've got them all and I'll look again for any that are missed.

(FWIW, it's not my preferred convention, so I spot these by looking for variable names that look 'right' to me, and know that they're actually 'wrong' because they look 'right' to me.)

File: clang/test/AST/HLSL/StructuredBuffer-AST.hlsl (hunk 7 above)
Comment (damyanp) [reply clang/test/AST/HLSL/StructuredBuffer-AST.hlsl:1]: We should try to fix this before completing the PR.

**Review by damyanp — COMMENTED** [review]
Some more notes

File: clang/lib/Sema/HLSLExternalSemaSource.cpp (hunk 11 above)
Comment (bob80905) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:343]: Yes, this was meant to show the AST structure.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp (hunk 6 above)
Comment (bob80905) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp]: Added! I've left out the expected AST structure since it probably won't ever conform to clang format.

File: clang/test/AST/HLSL/StructuredBuffer-AST.hlsl (hunk 7 above)
Comment (bob80905) [reply clang/test/AST/HLSL/StructuredBuffer-AST.hlsl]: Fixed!

File: clang/lib/Sema/HLSLExternalSemaSource.cpp (hunk 14 above)
Comment (bob80905) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp]: I think I've got them all.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp (hunk 12 above)
Comment (damyanp) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp]: There's some remaining in getSizeOfLEQ16Expr if you wanted to also update that.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp (hunk 14 above)
Comment (damyanp) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp]: I couldn't find any more!

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 15]
```diff
@@ -323,30 +324,128 @@ struct TemplateParameterListBuilder {
...
+
+  int fn() {
+      RWBuffer<int> Buf;
+  }
+
+  When dumping the AST and filtering for "RWBuffer", the resulting AST
+  structure is what we're trying to construct below, specifically the
+  CSE portion.
+  */
+  ConceptSpecializationExpr *getConceptSpecializationExpr(Sema &S,
```
Comment (damyanp) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:361]: nit on naming....

`getConceptSpecializationExpr` suggests to me that this function will look up some existing thing and return in. When in actual fact it is creating a bunch of things.

Other functions in this file use names more like `addConceptSpecializationExpr` or `defineConceptSpecializationExpr`.

This comment applies to pretty much all of the `getXXX` functions added in this PR.

**Review by damyanp — COMMENTED** [review]
LGTM. Would be good to get review from someone more familiar with ASTs and external sema source.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp (hunk 3 above)
Comment (hekota) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:334]: `referenced' means the declaration is referenced/used somewhere in the source code, you should not be setting it here.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 16]
```diff
@@ -483,10 +582,101 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
...
+
+  // Add the concept declaration to the Translation Unit Decl
+  Context.getTranslationUnitDecl()->addDecl(CD);
+
+  return CD;
+}
+
 void HLSLExternalSemaSource::defineHLSLTypesWithForwardDeclarations() {
   CXXRecordDecl *Decl;
+  ConceptDecl *CD = constructTypedBufferConceptDecl(*SemaPtr);
```
Comment (hekota) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:676]: I'd probably use a more descriptive name here, like `TypeBufferConcept` since as we'll need to have one for RawBuffers as well, right?

Comment (bob80905) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp]: Right, updated.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 17]
```diff
@@ -483,10 +581,102 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
...
+  return binaryOperator;
+}
+
+Expr *constructTypedBufferConstraintExpr(Sema &S, SourceLocation NameLoc,
+                                         TemplateTypeParmDecl *T) {
+  ASTContext &Context = S.getASTContext();
+
+  // first get the "sizeof(T) <= 16" expression, as a binary operator
+  BinaryOperator *SizeOfLEQ16 = constructSizeOfLEQ16Expr(Context, NameLoc, T);
+  // TODO: add the '__builtin_hlsl_is_line_vector_layout_compatible' builtin
```
Comment (llvm-beanz) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:627]: I think we had agreed on a different name for this. Also is there an issue filed that you can reference in this TODO?

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 18]
```diff
@@ -483,10 +581,102 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
...
+  // and return a binary operator that evaluates the builtin on the given
+  // template type parameter 'T'
+  return SizeOfLEQ16;
+}
+
+ConceptDecl *constructTypedBufferConceptDecl(Sema &S) {
+  DeclContext *DC = S.CurContext;
```
Comment (llvm-beanz) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:634]: Do we want this to be Sema's current context, or the HLSL namespace context? My guess is the later.
It probably doesn't matter too much, but it would be nice to make sure we're not injecting AST nodes into the top-level declaration context or a user-defined context.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 19]
```diff
@@ -483,10 +581,102 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
...
+  ASTContext &Context = S.getASTContext();
+  SourceLocation DeclLoc = SourceLocation();
+
+  IdentifierInfo &IsValidLineVectorII =
+      Context.Idents.get("is_valid_line_vector");
```
Comment (llvm-beanz) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:639]: I think we also had a different name for this...
Can you also prefix the name with double underscore (`__`) since it is an implementation detail in the compiler rather than an intentionally user-surfaced language feature?

**Review by llvm-beanz — COMMENTED** [review]
Mostly looking good to me. A few small comments.

**Comment by hekota** [comment]
Since your PR that adds `__builtin_hlsl_is_typed_resource_element_compatible` is merged do you want to incorporate here?

**Comment by bob80905** [comment]
> Since your PR that adds `__builtin_hlsl_is_typed_resource_element_compatible` is merged do you want to incorporate here?

I'd prefer not to, because this PR is already pretty big and I have a separate task that is singularly dedicated to finalizing the constraint expression, which will include incorporating the new builtin.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 20]
```diff
@@ -483,10 +581,105 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
...
+
+  // Add the concept declaration to the Translation Unit Decl
+  Context.getTranslationUnitDecl()->addDecl(CD);
+
+  return CD;
+}
+
 void HLSLExternalSemaSource::defineHLSLTypesWithForwardDeclarations() {
   CXXRecordDecl *Decl;
+  ConceptDecl *TypeBufferConcept =
```
Comment (hekota) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:677]: Typo
```suggestion
  ConceptDecl *TypedBufferConcept =
```

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 21]
```diff
@@ -323,30 +324,127 @@ struct TemplateParameterListBuilder {
...
-  BuiltinTypeDeclBuilder &finalizeTemplateArgs() {
+  /*
+  The concept specialization expression (CSE) constructed below is constructed
+  so that it matches the CSE that is constructed when parsing
+  the below C++ code:
+
+  template<typename T>
+  concept is_valid_line_vector =sizeof(T) <= 16;
+
+  template<typename element_type> requires is_valid_line_vector<element_type>
```
Comment (hekota) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:346]: This still uses the old name.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 22]
```diff
@@ -323,30 +324,127 @@ struct TemplateParameterListBuilder {
...
+        /*depth=*/0,                // Depth in the template parameter list
+        /*position=*/0,             // Position in the template parameter list
+        /*id=*/nullptr,             // Identifier for 'T'
+        /*Typename=*/true,          // Indicates this is a 'typename' or 'class'
+        /*ParameterPack=*/false,    // Not a parameter pack
+        /*HasTypeConstraint=*/false // Has no type constraint
+    );
+
+    T->setDeclContext(DC);
+    T->setReferenced();
```
Comment (hekota) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:392]: Why do you need to set this explicitly?

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 23]
```diff
@@ -483,10 +581,105 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
...
+  return SizeOfLEQ16;
+}
+
+ConceptDecl *constructTypedBufferConceptDecl(Sema &S, NamespaceDecl *NSD) {
+  ASTContext &Context = S.getASTContext();
+  DeclContext *DC = NSD->getDeclContext();
+  SourceLocation DeclLoc = SourceLocation();
+
+  IdentifierInfo &IsTypedResourceElementCompatibleII =
+      Context.Idents.get("__is_typed_resource_element_compatible");
```
Comment (hekota) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:640]: Nit - move this just before DeclName declaration, or better yet use `Context.Idents.get("__is_typed_resource_element_compatible")` directly in the DeclarationName constructor.
It makes it easier to read when variables are declared closer to where they are is used.

**Review by hekota — APPROVED** [review]
Few minor things, otherwise LGTM!

File: clang/lib/Sema/HLSLExternalSemaSource.cpp (hunk 22 above)
Comment (bob80905) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:392]: Setting the decl context is necessary because it puts the decl into the right indentation in the AST dump. Otherwise, I believe the decl would be placed at the same scope as one indentation below the translation unit decl, which is not where the decl belongs.
I'll see if I can get away with removing setReferenced here.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 24]
```diff
@@ -323,30 +324,127 @@ struct TemplateParameterListBuilder {
...
+  The concept specialization expression (CSE) constructed below is constructed
+  so that it matches the CSE that is constructed when parsing
+  the below C++ code:
+
+  template<typename T>
+  concept is_typed_resource_element_compatible =sizeof(T) <= 16;
+
+  template<typename element_type> requires
+  is_typed_resource_element_compatible<element_type>
+
```
Comment (damyanp) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:348]: The blank line between the template bit and the struct really confused me for a while trying to read this.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 25]
```diff
@@ -323,30 +324,127 @@ struct TemplateParameterListBuilder {
...
+  struct RWBuffer {
+      element_type Val;
```
Comment (damyanp) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:350]: I don't think that the code in this comment really matches the AST that's being built.

RWBuffer doesn't have a member of type element_type, for example.  

I _think_ that this function is just building up the AST that corresponds to the `requires is_typed_resource_element_compatible<element_Type>` part?

The AST nodes for ` template<typename T> concept is_typed_resource_element_compatible =sizeof(T) <= 16;`
itself is created in constructTypedBufferConceptDecl?

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 26]
```diff
@@ -323,30 +324,127 @@ struct TemplateParameterListBuilder {
...
+        /*Typename=*/true,          // Indicates this is a 'typename' or 'class'
+        /*ParameterPack=*/false,    // Not a parameter pack
+        /*HasTypeConstraint=*/false // Has no type constraint
+    );
+
+    T->setDeclContext(DC);
+
+    QualType ConceptTType = Context.getTypeDeclType(ConceptTTPD);
+
+    // this is the 2nd template argument node in the AST above
```
Comment (damyanp) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:396]: I'm not entirely sure what "the AST above" refers to.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 27]
```diff
@@ -323,30 +324,127 @@ struct TemplateParameterListBuilder {
...
+    // Assume that the concept decl has just one template parameter
+    // This parameter should have been added when CD was constructed
+    // in getTypedBufferConceptDecl
+    assert(CD->getTemplateParameters()->size() == 1 &&
+           "unexpected concept decl parameter count");
+    TemplateTypeParmDecl *ConceptTTPD = dyn_cast<TemplateTypeParmDecl>(
+        CD->getTemplateParameters()->getParam(0));
+
+    // this fake TemplateTypeParmDecl is used to construct a template argument
+    // that will be used to construct the ImplicitConceptSpecializationDecl
```
Comment (damyanp) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:379]: What's fake about it?

**Review by damyanp — COMMENTED** [review]
Some notes, mostly about the comments.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp (hunk 26 above)
Comment (bob80905) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:396]: Ah, this was when I had the entire AST inside a comment previously, haven't updated it, will fix!

File: clang/lib/Sema/HLSLExternalSemaSource.cpp (hunk 27 above)
Comment (bob80905) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:379]: It isn't being placed into the AST, the sole reason for its existence is to allow a path to construct a TemplateArgument, and that TemplateArgument will actually be placed into the AST. The fake TemplateTypeParmDecl is otherwise unused and discarded.

Comment (llvm-beanz) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:378]: This sounds wrong... Like, completely wrong. You're absolutely adding it to the AST, that's what happens when you create a declaration and put it into a declaration context.

This also seems like we're putting this declaration into the wrong declaration context.

I don't see any of your tests verifying the AST shape of the concept declaration. Can you add tests for that so that we can see what the concept's AST looks like?

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 28]
```diff
@@ -323,30 +324,129 @@ struct TemplateParameterListBuilder {
...
+           "unexpected concept decl parameter count");
+    TemplateTypeParmDecl *ConceptTTPD = dyn_cast<TemplateTypeParmDecl>(
+        CD->getTemplateParameters()->getParam(0));
+
+    // this TemplateTypeParmDecl is the template for the resource, and is
+    // used to construct a template argumentthat will be used
+    // to construct the ImplicitConceptSpecializationDecl
+    TemplateTypeParmDecl *T = TemplateTypeParmDecl::Create(
+        Context,                          // AST context
+        Context.getTranslationUnitDecl(), // DeclContext
```
Comment (llvm-beanz) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:382]: This is where the DeclContext is wrong that I referenced in https://github.com/llvm/llvm-project/pull/112600/files/6edf031b5e736b38cf3551ccc872351f9c8a07ca#r1835111011
```suggestion
        Builder.Record->getDeclContext();, // DeclContext
```

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 29]
```diff
@@ -483,10 +583,103 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
...
+}
+
+ConceptDecl *constructTypedBufferConceptDecl(Sema &S, NamespaceDecl *NSD) {
+  ASTContext &Context = S.getASTContext();
+  DeclContext *DC = NSD->getDeclContext();
+  SourceLocation DeclLoc = SourceLocation();
+
+  IdentifierInfo &ElementTypeII = Context.Idents.get("element_type");
+  TemplateTypeParmDecl *T = TemplateTypeParmDecl::Create(
+      Context, Context.getTranslationUnitDecl(), DeclLoc, DeclLoc,
```
Comment (llvm-beanz) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:643]: Pretty sure this is also wrong here:
```suggestion
      Context,  NSD->getDeclContext(), DeclLoc, DeclLoc,
```

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 30]
```diff
@@ -483,10 +583,103 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
...
+  TemplateParameterList *ConceptParams = TemplateParameterList::Create(
+      Context, DeclLoc, DeclLoc, {T}, DeclLoc, nullptr);
+
+  DeclarationName DeclName = DeclarationName(
+      &Context.Idents.get("__is_typed_resource_element_compatible"));
+  Expr *ConstraintExpr = constructTypedBufferConstraintExpr(S, DeclLoc, T);
+
+  // Create a ConceptDecl
+  ConceptDecl *CD =
+      ConceptDecl::Create(Context, Context.getTranslationUnitDecl(), DeclLoc,
```
Comment (llvm-beanz) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:663]:
```suggestion
      ConceptDecl::Create(Context, NSD->getDeclContext(), DeclLoc,
```

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 31]
```diff
@@ -483,10 +583,103 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
...
+                          DeclName, ConceptParams, ConstraintExpr);
+
+  // Attach the template parameter list to the ConceptDecl
+  CD->setTemplateParameters(ConceptParams);
+
+  // Add the concept declaration to the Translation Unit Decl
+  Context.getTranslationUnitDecl()->addDecl(CD);
```
Comment (llvm-beanz) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:670]: We should be putting the concepts under the `hlsl` namespace not under the top level declaration where they may conflict with user-defined declarations.

```suggestion
  NSD->getDeclContext()->addDecl(CD);
```

**Review by llvm-beanz — APPROVED** [review]
